File: DSA/DSA_450/Cpp/10_stack_and_queues/029_Interleave_Queue.cpp
/*
Problem: Interleave First Half of Queue with Second Half
URL: https://www.geeksforgeeks.org/interleave-first-half-queue-second-half/

Problem Statement:
Given a queue of even size, interleave first half with second half.

Sample Input/Output:
Input: [11,12,13,14,15,16,17,18,19,20]
Output: [11,16,12,17,13,18,14,19,15,20]
*/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    queue<int> Interleave_Queue_Stack(queue<int>& q) {
        int n = q.size();
        int half = n / 2;
        stack<int> st;
        
        for (int i = 0; i < half; i++) {
            st.push(q.front());
            q.pop();
        }
        
        while (!st.empty()) {
            q.push(st.top());
            st.pop();
        }
        
        for (int i = 0; i < half; i++) {
            q.push(q.front());
            q.pop();
        }
        
        for (int i = 0; i < half; i++) {
            st.push(q.front());
            q.pop();
        }
        
        while (!st.empty()) {
            q.push(st.top());
            st.pop();
            q.push(q.front());
            q.pop();
        }
        
        return q;
    }

    queue<int> Interleave_Queue_Auxiliary(queue<int>& q) {
        int n = q.size();
        int half = n / 2;
        queue<int> aux;
        
        for (int i = 0; i < half; i++) {
            aux.push(q.front());
            q.pop();
        }
        
        while (!aux.empty()) {
            q.push(aux.front());
            aux.pop();
            q.push(q.front());
            q.pop();
        }
        
        return q;
    }
};

void Test_Interleave_Queue_Stack() {
    Solution solution;
    
    queue<int> q1;
    q1.push(11);
    q1.push(12);
    q1.push(13);
    q1.push(14);
    q1.push(15);
    q1.push(16);
    q1.push(17);
    q1.push(18);
    q1.push(19);
    q1.push(20);
    
    queue<int> result1 = solution.Interleave_Queue_Stack(q1);
    cout << "Stack - Interleaved Queue: ";
    while (!result1.empty()) {
        cout << result1.front() << " ";
        result1.pop();
    }
    cout << endl;
}

void Test_Interleave_Queue_Auxiliary() {
    Solution solution;
    
    queue<int> q2;
    q2.push(1);
    q2.push(2);
    q2.push(3);
    q2.push(4);
    q2.push(5);
    q2.push(6);
    
    queue<int> result2 = solution.Interleave_Queue_Auxiliary(q2);
    cout << "Auxiliary - Interleaved Queue: ";
    while (!result2.empty()) {
        cout << result2.front() << " ";
        result2.pop();
    }
    cout << endl;
}

int main() {
    Test_Interleave_Queue_Stack();
    Test_Interleave_Queue_Auxiliary();
    return 0;
}

File: DSA/DSA_450/Python/10_stack_and_queues/029_Interleave_Queue.py
"""
Problem: Interleave First Half of Queue with Second Half
URL: https://www.geeksforgeeks.org/interleave-first-half-queue-second-half/

Problem Statement:
Given a queue of even size, interleave first half with second half.

Sample Input/Output:
Input: [11,12,13,14,15,16,17,18,19,20]
Output: [11,16,12,17,13,18,14,19,15,20]
"""

from collections import deque


class Solution:
    def Interleave_Queue_Stack(self, q):
        """
        Interleave queue using stack.
        Time Complexity: O(n)
        Space Complexity: O(n)
        """
        n = len(q)
        half = n // 2
        st = []
        
        for i in range(half):
            st.append(q.popleft())
        
        while st:
            q.append(st.pop())
        
        for i in range(half):
            q.append(q.popleft())
        
        for i in range(half):
            st.append(q.popleft())
        
        while st:
            q.append(st.pop())
            q.append(q.popleft())
        
        return q

    def Interleave_Queue_Auxiliary(self, q):
        """
        Interleave queue using auxiliary queue.
        Time Complexity: O(n)
        Space Complexity: O(n)
        """
        n = len(q)
        half = n // 2
        aux = deque()
        
        for i in range(half):
            aux.append(q.popleft())
        
        while aux:
            q.append(aux.popleft())
            q.append(q.popleft())
        
        return q


def Test_Interleave_Queue_Stack():
    solution = Solution()
    
    q1 = deque([11, 12, 13, 14, 15, 16, 17, 18, 19, 20])
    result1 = solution.Interleave_Queue_Stack(q1)
    print("Stack - Interleaved Queue: ", end="")
    while result1:
        print(result1.popleft(), end=" ")
    print()


def Test_Interleave_Queue_Auxiliary():
    solution = Solution()
    
    q2 = deque([1, 2, 3, 4, 5, 6])
    result2 = solution.Interleave_Queue_Auxiliary(q2)
    print("Auxiliary - Interleaved Queue: ", end="")
    while result2:
        print(result2.popleft(), end=" ")
    print()


if __name__ == "__main__":
    Test_Interleave_Queue_Stack()
    Test_Interleave_Queue_Auxiliary()
